### DISCUSSION TITLE
Make Minio accessible behind a Cloudflare proxy

### DISCUSSION THREAD

**Discussion by mvartuc** [minio/minio#20188] [Q&A] [answered]
As explained here: #15506  ,  a signature mismatch error currently occurs when accessing minio behind a proxy. I'm currently trying to self host everything for my web app and minio, for some reason, is the only service that refuses to work under a cloudflare tunnel even though I disabled caching.

Can this not be worked on so that people like me who have their local internet behind a CGNAT can self host?

**Comment by yaozhenghangma** [comment]
I also use cloudlfared tunnel. The service works when I disable caching. I test the connection using S3 File and S3Drive, and they all work well.

**Comment by harshavardhana** [comment, marked as answer]
CloudFlare caches response causing signature issues, disable their caching.

**Comment by andrekutianski** [comment]
For record, a good reading to help with that https://www.linkedin.com/pulse/resolving-minio-behind-cloudflare-proxy-403-error-bhavesh-deshmukh-pk1nf/

**Comment by speedrapide10** [comment]
in my case issue still persist even after i set up the cache bypass rule :
![image](https://github.com/user-attachments/assets/e67427c1-3ead-42bf-85b0-1cb6db10b692)

The bucket is accessible properly without cloudflare tunnel. any suggestions ?

Thank you

**Reply by speedrapide10** [reply]
i managed to fix the issue by using my nginx proxy instead of cloudflare tunnel

**Reply by andrekutianski** [reply]
Are you using a domain like `my.s3.domain.tld` in Minio? Cloudflare cannot issue certificates for subdomains of subdomains.

**Reply by speedrapide10** [reply]
no i'm using s3.domain.tld

**Comment by wlvpaas** [comment]
I ended up discovering by chance that Cloudflare manipulates some headers that hinder the functioning of Minio. For me, the big villain is the `Add security headers` of the 'HTTP response headers' setting available under the 'Managed Transformers'. This setting adds several security-related HTTP response headers providing cross-site scripting (XSS) protection but breaks the Minio dashboard because it prevents the browser from interpreting different content types that are advertised in the headers. 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options

**Reply by speedrapide10** [reply]
Thank you all for your help 💪🏻. i'll make sure to test it later. however i think i will stick to my selfhosted reverse proxy to prevent any security issue for my cloudflare tunnel.

**Reply by kaido2ki** [reply]
did you fix it?

**Reply by speedrapide10** [reply]
@kaido2ki Yes i ended up using my selfhosted reverse proxy instead of cloudflare tunnel to fix this issue.
And sorry for the late reply

**Reply by mobilestack** [reply]
> I ended up discovering by chance that Cloudflare manipulates some headers that hinder the functioning of Minio. For me, the big villain is the `Add security headers` of the 'HTTP response headers' setting available under the 'Managed Transformers'. This setting adds several security-related HTTP response headers providing cross-site scripting (XSS) protection but breaks the Minio dashboard because it prevents the browser from interpreting different content types that are advertised in the headers.
> 
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options

and finally, how to solve the problem?

**Comment by devkamiki** [comment]
Hello,
I have the same problem too, and couldn't be fixed with disabling caching.
( When I was writing this comment I didn't test mc so I thought it's header thing and won't work anywhere but surprisingly it just doesn't work w/ rclone, which would response as signature doesn't match, but worked in mc.)
In my case, I tried to use the same credentials in rclone and mc, rclone remote named `minio` and mc set the endpoint url's alias `one` .
```
[yuki@manjaro ~]$ rclone ls minio:              
2025/04/02 21:47:01 NOTICE: Failed to ls: operation error S3: ListBuckets, https response error StatusCode: 403, RequestID: <requestid>, HostID: <hostid>, api error SignatureDoesNotMatch: The request signature we calculated does not match the signature you provided. Check your key and signing method.                                                                              
[yuki@manjaro ~]$ mc ls one    
[2025-04-02 15:42:43 CST]     0B test/      
[yuki@manjaro ~]$ mc ls one/test                  
[2025-04-02 16:40:17 CST]  33KiB STANDARD <picturename>.jpg  
```
It's probably more like rclone problem for me, I'm searching similar issues from rclone repo

**Reply by maxpain** [reply]
rclone uses `aws-sdk-go-v2` library. I wrote a Golang problem using this library and got the same error.

**Reply by ken-kuro** [reply]
The beta version of rclone has implemented a [workaround](https://github.com/rclone/rclone/commit/e175c863aa9950b67cebd18d95c56ed547a012f7) for this issue by introducing middleware that removes and re-adds the relevant header. You can leverage the new `sign_accept_encoding` parameter to enable MinIO to function correctly behind a Cloudflare proxy. Although this parameter was originally intended to address issues with GCS, it is applicable here as both services attempt to modify request headers in a similar manner.

As for `aws-sdk-go-v2`, it seems unlikely that AWS will implement an upstream fix for this in the near future.

**Comment by richardARPANET** [comment]
Solution = make a cache rule like below. 

note my minio hostname is object-storage.example.com

<img width="601" alt="image" src="https://github.com/user-attachments/assets/166e88bd-e53a-4dfa-9229-007f33176096" />

**Reply by ken-kuro** [reply]
Bypass Cache rule is required for MinIO to work, but CF still messes with header like `accept-encoding` causing mismatch singature for tools like RClone, do you know how to prevent that ?

**Reply by richardARPANET** [reply]
Not sure as I'm not using rclone, however awscli does work with the above.

**Reply by ken-kuro** [reply]
I see. Tool like rclone actually use aws sdk (go v2 specifically) to interact with the MinIO, and due to the differences between SDKs, some has signed header which doesn't work with CF (go v2 has that problem, while latest js v3 doesn't)
